### PR TITLE
Move disclosed contract logic to speedy

### LINKED ISSUE #14090: Handle disclosed contracts in speedy instead of Daml Engine
There's quite a bit of duplicated logic between the `ScenarioRunner` and `speedy.Engine` that should be reduced.

### PR DESCRIPTION
**Pull request by robin-da**
This moves the disclosed contract logic from the engine to speedy.

Fixes #14090.

### CODE REVIEW

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala [hunk 1]
```diff
@@ -1071,24 +1070,32 @@ private[lf] object SBuiltin {
...
-                  // The call to ToCachedContractDefRef(actualTmplId) will query package
-                  // of actualTmplId if not know.
-                  SEVal(ToCachedContractDefRef(actualTmplId)),
-                  Array(
-                    SEImportValue(Ast.TTyCon(actualTmplId), arg),
-                    SEValue(args.get(1)),
-                  ),
-                )
-              },
+          def continue = { case V.ContractInstance(actualTmplId, arg, _) =>
```
Comment (cocreature) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:1073]: if you want to simplify things in Speedy, consider having `ResultNeedContract` keep a versioned contract instance but only drop it from `SResultNeedContract`. Then you don’t break Canton & the participant.

**Review by cocreature — APPROVED** [review]
This looks sensible. I suggest undoing the change to ResultNeedContract but keep SResultNeedContract if you want to. That reduces unnecessary API breakage.

We also really need some Engine or speedy-level tests for this but if you want, I’m fine with leaving that for a separate PR.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala (hunk 1 above)
Comment (robin-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:1073]: Ok, I did that.

**Comment by robin-da** [comment]
@carlpulley-da I created a ticket to add tests to this.

**Pull request merged**